api/src/main/java/org/eclipse/microprofile/reactive/streams/Completion.java

```diff
@@ -0,0 +1,33 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package org.eclipse.microprofile.reactive.streams;
+
+/**
+ * A stream that completes with a single result.
+ * <p>
+ * This will either be a {@link SubscriberBuilder}, representing a stream with an inlet that can be plumbed to a
+ * publisher in order to run it, or a {@link CompletionRunner}, representing a closed graph that can be run as is.
+ *
+ * @param <T> The result of the stream.
+ * @see SubscriberBuilder
+ * @see CompletionRunner
+ */
+public interface Completion<T> {
```

Comment:
Do we want to have the <T> param here ?

Reply:
I don't think it's technically needed, but without it, `CompletionOperators` would be full of methods like this:

```java
Completion toList();
Completion findFirst();
Completion ignore();
```

One of the reasons for splitting the interfaces out is grouping documentation, but looking at those motheds, what do those type signatures say about what the above methods return? Nothing. By having the parameter, we can show that in the method signature:

```java
Completion<List<T>> toList();
Completion<Optional<T>> findFirst();
Completion<Void> ignore();
```

And we can also enforce consistency in return types of `PublisherBuilder` and `ProcessorBuilder` using the type system - if `PublisherBuilder` tried to return `CompletionRunner<Void>` from `ignore()`, while `ProcessorBuilder` tried to return `SubscriberBuilder<R, R>` instead of `SubscriberBuilder<R, Void>` as it should be, if `Completion` didn't have a type parameter, that would be allowed, but since it does have a type parameter, trying to do the above would fail to compile.